src/test/java/org/opensearch/security/dlic/dlsfls/FlsFlatTests.java

```diff
@@ -0,0 +1,246 @@
+import org.opensearch.common.xcontent.XContentFactory;
+import org.opensearch.security.test.DynamicSecurityConfig;
+import org.opensearch.security.test.helper.rest.RestHelper.HttpResponse;
+import org.opensearch.core.xcontent.XContentBuilder;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.core.IsEqual.equalTo;
+import static org.hamcrest.core.IsNot.not;
+import static org.hamcrest.core.StringContains.containsString;
+
+public class FlsFlatTests extends AbstractDlsFlsTest {
+
+    private final String ROLES_FILE = "roles_fls_flat.yml";
+    private final String ROLES_MAPPINGS_FILE = "roles_mapping_fls_indexing.yml";
+
+    protected void populateData(final Client tc) {
+        // Create several documents in different indices with shared field names,
+        // different roles will have different levels of FLS restrictions
+
+        final BiFunction<XContentBuilder, String, XContentBuilder> addFlatField = (builder, fieldName) -> {
+            try {
+                return builder.startObject(fieldName)
+                    .startObject("properties")
+                    .startObject("field")
+                    .field("type", "flat_object")
+                    .endObject()
+                    .endObject()
+                    .endObject();
+            } catch (final Exception e) {
+                throw new RuntimeException(e);
+            }
+        };
+
+        XContentBuilder builder = null;
+        try {
+            builder = XContentFactory.jsonBuilder().startObject();
+            builder = addFlatField.apply(builder, "phone-all");
+            builder = addFlatField.apply(builder, "phone-some");
+            builder = addFlatField.apply(builder, "phone-one");
+            builder.endObject();
+        } catch (final Exception e) {
+            throw new RuntimeException(e);
+        }
+        final String mappings = Strings.toString(builder);
+
+        final Consumer<String> createIndexWithMapping = (indexName) -> {
+            final CreateIndexRequest createIndex = new CreateIndexRequest(indexName);
+            createIndex.settings(mappings, XContentType.JSON);
+            tc.admin().indices().create(createIndex);
+        };
+
+        // Field Schema
+        // - phone-all.areaCode: {docId}001
+        // - phone-some.areaCode: {docId}002
+        // - phone-one.areaCode: {docId}002
+        // Local number == areaCode + 90
+        // Filtering is only done on local number
+        createIndexWithMapping.accept("yellow-page");
+        tc.index(
+            new IndexRequest("yellow-pages").id("1")
+                .setRefreshPolicy(RefreshPolicy.IMMEDIATE)
+                .source(
+                    "{\"phone-all\": {\"areaCode\": 1001, \"localNumber\": 1091 },\"phone-some\":{\"areaCode\": 1002, \"localNumber\": 1092 },\"phone-one\":{\"areaCode\": 1003, \"localNumber\": 1093 }}",
```

Comment:
Nit: For me personally, it is difficult to read such code. What do you think about using a JSON object builder, e.g., `XContentBuilder` to build a JSON payload?